powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/util/TokenUtils.java

```diff
@@ -92,29 +92,33 @@ public byte[] convertTokenTimestamp(long timestamp) {
      * Compute the digest of provided token information using given token secret.
      * @param nonce Token nonce, 16 random bytes.
      * @param timestamp Token timestamp, Unix timestamp format encoded as bytes (string representation).
+     * @param version Protocol version.
      * @param tokenSecret Token secret, 16 random bytes.
      * @return Token digest computed using provided data bytes with given token secret.
      * @throws GenericCryptoException In case digest computation fails.
      * @throws CryptoProviderException In case cryptography provider is incorrectly initialized.
      */
-    public byte[] computeTokenDigest(byte[] nonce, byte[] timestamp, byte[] tokenSecret) throws GenericCryptoException, CryptoProviderException {
-        byte[] amp = "&".getBytes(StandardCharsets.UTF_8);
-        byte[] data = ByteUtils.concat(nonce, amp, timestamp);
+    public byte[] computeTokenDigest(byte[] nonce, byte[] timestamp, byte[] version, byte[] tokenSecret) throws GenericCryptoException, CryptoProviderException {
+        final byte[] amp = "&".getBytes(StandardCharsets.UTF_8);
+        final byte[] data = (version != null)
+                ? ByteUtils.concat(nonce, amp, timestamp, amp, version)
+                : ByteUtils.concat(nonce, amp, timestamp);
```

Comment:
I think that we should make version mandatory parameter and the logic whether we include it to calculation or not should be implemented here. Like  `if ("3.2".equals(version)) {...}`. Otherwise it will be yet another example where we delegate the protocol related logic to outside of the library. The result is a lot more spaghetti copy-pasted code in three (or more) repositories.

Reply:
At least we should do this logic in `ClientTokenGenerator` or `ServerTokenVerifier` classes.